src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -250,28 +250,46 @@ private static boolean containsExcludedFlags(String testStr) {
        return false;
     }
 
+    // Return a temp file name if the archive file exists. The temp file operation may
+    // cause permission exception which prevents real archive being created.
+    private static String getArchiveFileName(String name) throws Exception {
+        String tempFileName = name + ".temp";
+        File tempFile = new File(tempFileName);
+        if (tempFile.exists()) {
+            tempFile.delete();
+        } else {
+            if (tempFile.createNewFile()) {
+                tempFile.delete();
+            }
+        }
+
+        File archiveFile = new File(name);
+        if (!archiveFile.exists()) {
+            return name;
+        }
+        return tempFileName;
+    }
+
     /**
     * called from jcmd VM.cds to dump static or dynamic shared archive
     * @param isStatic true for dump static archive or false for dynnamic archive.
     * @param fileName user input archive name, can be null.
     */
     private static void dumpSharedArchive(boolean isStatic, String fileName) throws Exception {
         String currentPid = String.valueOf(ProcessHandle.current().pid());
-        String archiveFile =  fileName != null ? fileName :
+        String archiveFileName =  fileName != null ? fileName :
             "java_pid" + currentPid + (isStatic ? "_static.jsa" : "_dynamic.jsa");
 
-        // delete if archive file aready exists
-        File fileArchive = new File(archiveFile);
-        if (fileArchive.exists()) {
-            fileArchive.delete();
-        }
+        String tempFileName = getArchiveFileName(archiveFileName);
+        File tempArchiveFile = new File(tempFileName);
+
         if (isStatic) {
-            String listFile = archiveFile + ".classlist";
-            File fileList = new File(listFile);
-            if (fileList.exists()) {
-                fileList.delete();
+            String listFileName = tempFileName + ".classlist";
+            File listFile = new File(listFileName);
```

Comment:
There's no need to use tempFileName for the classlist. The list file is always deleted after the dump has finished.

Reply:
Yes, will revert it back.